File: src/workspace_workbench/core/git.py
from __future__ import annotations

from dataclasses import dataclass
import hashlib
import os
from pathlib import Path
import subprocess
from typing import Iterable, Sequence

from .errors import GitCommandError, WorkbenchError


@dataclass(frozen=True)
class GitResult:
    args: tuple[str, ...]
    returncode: int
    stdout: str
    stderr: str


@dataclass(frozen=True)
class GitFile:
    path: str
    status: str
    old_path: str | None = None
    additions: int | None = None
    deletions: int | None = None
    binary: bool = False


def _decode(value: bytes) -> str:
    return value.decode("utf-8", errors="replace")


class GitClient:
    """Small, bounded, read-mostly wrapper around the Git CLI.

    No method accepts a shell string. Arguments are passed directly to Git so
    a project config or UI cannot turn an observation into arbitrary command
    execution.
    """

    def __init__(self, repository: str | Path, *, timeout: float = 3.0) -> None:
        self.repository = Path(repository).expanduser().resolve()
        self.timeout = max(0.5, float(timeout))

    def run(
        self,
        args: Sequence[str],
        *,
        timeout: float | None = None,
        check: bool = True,
    ) -> GitResult:
        normalized = tuple(str(value) for value in args)
        try:
            completed = subprocess.run(
                ["git", "-c", "core.fsmonitor=false", "-C", str(self.repository), *normalized],
                stdout=subprocess.PIPE,
                stderr=subprocess.PIPE,
                check=False,
                timeout=timeout if timeout is not None else self.timeout,
                env={**os.environ, "GIT_OPTIONAL_LOCKS": "0", "GIT_TERMINAL_PROMPT": "0"},
            )
        except subprocess.TimeoutExpired as exc:
            raise GitCommandError(
                f"git command timed out after {timeout if timeout is not None else self.timeout:g}s",
                code="git_timeout",
                details={"repository": str(self.repository), "args": list(normalized)},
            ) from exc
        except OSError as exc:
            raise GitCommandError(
                f"git command could not start: {exc}",
                code="git_unavailable",
                details={"repository": str(self.repository)},
            ) from exc
        result = GitResult(normalized, completed.returncode, _decode(completed.stdout), _decode(completed.stderr))
        if check and result.returncode != 0:
            detail = result.stderr.strip() or result.stdout.strip() or "git command failed"
            raise GitCommandError(
                detail,
                code="git_command_failed",
                details={"repository": str(self.repository), "args": list(normalized), "returncode": result.returncode},
            )
        return result

    def is_repository(self) -> bool:
        if not self.repository.is_dir():
            return False
        result = self.run(["rev-parse", "--is-inside-work-tree"], check=False)
        return result.returncode == 0 and result.stdout.strip() == "true"

    def root(self) -> Path:
        value = self.run(["rev-parse", "--show-toplevel"]).stdout.strip()
        return Path(value).resolve()

    def head(self) -> str | None:
        result = self.run(["rev-parse", "--verify", "HEAD"], check=False)
        return result.stdout.strip() if result.returncode == 0 else None

    def branch(self) -> str | None:
        result = self.run(["symbolic-ref", "--quiet", "--short", "HEAD"], check=False)
        value = result.stdout.strip()
        return value or None

    def upstream(self) -> tuple[str | None, str | None]:
        name = self.run(["rev-parse", "--abbrev-ref", "--symbolic-full-name", "@{upstream}"], check=False).stdout.strip()
        sha = self.run(["rev-parse", "--verify", "@{upstream}"], check=False).stdout.strip()
        return name or None, sha or None

    def refs(self, *, head: str | None = None) -> list[dict[str, str | bool]]:
        args = ["for-each-ref"]
        if head:
            args.append(f"--merged={head}")
        args.extend(["--format=%(objectname)\t%(refname)\t%(symref)\t%(*objectname)", "refs/heads", "refs/remotes", "refs/tags"])
        output = self.run(args).stdout
        result: list[dict[str, str | bool]] = []
        for line in output.splitlines():
            parts = line.split("\t")
            if len(parts) < 2:
                continue
            sha, name = parts[0].strip(), parts[1].strip()
            if not sha or not name:
                continue
            short_name = name.removeprefix("refs/heads/").removeprefix("refs/remotes/").removeprefix("refs/tags/")
            kind = "tag" if name.startswith("refs/tags/") else "remote" if name.startswith("refs/remotes/") else "local"
            if kind == "tag" and len(parts) > 3 and parts[3].strip():
                sha = parts[3].strip()
            result.append({"name": name, "shortName": short_name, "sha": sha, "kind": kind})
        result.sort(key=lambda item: (str(item["kind"]), str(item["shortName"])))
        return result

    def status_porcelain(self) -> list[tuple[str, str]]:
        output = self.run(["status", "--porcelain=v1", "-z", "--untracked-files=all"]).stdout
        values = output.split("\x00")
        result: list[tuple[str, str]] = []
        index = 0
        while index < len(values):
            value = values[index]
            index += 1
            if not value:
                continue
            status = value[:2]
            path = value[3:] if len(value) > 3 else ""
            if status and status[0] in {"R", "C"} and index < len(values) and values[index]:
                old_path = path
                path = values[index]
                index += 1
                result.append((status, f"{old_path}\x00{path}"))
            elif path:
                result.append((status, path))
        return result

    def _numstat(self, args: Sequence[str]) -> dict[str, tuple[int | None, int | None, bool]]:
        output = self.run([*args, "--numstat", "-z"]).stdout
        result: dict[str, tuple[int | None, int | None, bool]] = {}
        values = output.split("\x00")
        index = 0
        while index < len(values):
            value = values[index]
            index += 1
            if not value:
                continue
            parts = value.split("\t", 2)
            if len(parts) != 3:
                continue
            additions_raw, deletions_raw, path = parts
            if not path and index + 1 < len(values):
                path = values[index + 1]
                index += 2
            try:
                additions = int(additions_raw)
                deletions = int(deletions_raw)
                binary = False
            except ValueError:
                additions = deletions = None
                binary = True
            result[path] = (additions, deletions, binary)
        return result

    def changed_files(self, scope: str, *, base: str | None = None, commit: str | None = None) -> list[GitFile]:
        if scope == "working":
            head = self.head()
            names = self.run(["diff", "--name-status", "-z", "--find-renames", "HEAD"], check=False) if head else GitResult((), 0, "", "")
            stats = self._numstat(["diff", "HEAD"]) if head else {}
            entries = self._parse_name_status(names.stdout)
            known = {entry.path for entry in entries}
            for status, path in self.status_porcelain():
                if status == "??" and path not in known:
                    entries.append(GitFile(path=path, status="A"))
                    known.add(path)
            for entry in entries:
                values = stats.get(entry.path)
                if values:
                    entry_index = entries.index(entry)
                    entries[entry_index] = GitFile(
                        path=entry.path,
                        status=entry.status,
                        old_path=entry.old_path,
                        additions=values[0],
                        deletions=values[1],
                        binary=values[2],
                    )
            return self._with_untracked_stats(entries)
        if scope == "commit":
            if not commit:
                raise WorkbenchError("commit is required", code="commit_required")
            commit = self.resolve_commit(commit)
            parents = self.run(["rev-list", "--parents", "-n", "1", commit]).stdout.split()[1:]
            args = ["diff", parents[0], commit] if parents else ["diff-tree", "--root", "--no-commit-id", "-r", commit]
            names = self.run([*args, "--name-status", "-z", "--find-renames"])
            stats = self._numstat(args)
        elif scope == "branch":
            if not base:
                raise WorkbenchError("base is required for branch changes", code="base_missing")
            head = self.head()
            if not head:
                return []
            names = self.run(["diff", "--name-status", "-z", "--find-renames", base, head])
            stats = self._numstat(["diff", base, head])
        else:
            raise WorkbenchError(f"unsupported change scope: {scope}", code="scope_invalid")
        entries = self._parse_name_status(names.stdout)
        return [
            GitFile(
                path=entry.path,
                status=entry.status,
                old_path=entry.old_path,
                additions=stats.get(entry.path, (None, None, False))[0],
                deletions=stats.get(entry.path, (None, None, False))[1],
                binary=stats.get(entry.path, (None, None, False))[2],
            )
            for entry in entries
        ]

    @staticmethod
    def _parse_name_status(output: str) -> list[GitFile]:
        values = output.split("\x00")
        result: list[GitFile] = []
        index = 0
        while index < len(values):
            value = values[index]
            index += 1
            if not value:
                continue
            if "\t" in value:
                raw_status, path = value.split("\t", 1)
            elif index < len(values):
                # `--name-status -z` uses NUL between the status and path,
                # unlike its line-oriented form which uses a tab.
                raw_status, path = value, values[index]
                index += 1
            else:
                continue
            status = raw_status[:1]
            old_path: str | None = None
            if status in {"R", "C"} and index < len(values):
                old_path, path = path, values[index]
                index += 1
            result.append(GitFile(path=path, status=status, old_path=old_path))
        return result

    def _with_untracked_stats(self, entries: list[GitFile]) -> list[GitFile]:
        result: list[GitFile] = []
        for entry in entries:
            if entry.status != "A" or entry.additions is not None or not (self.repository / entry.path).is_file():
                result.append(entry)
                continue
            try:
                candidate = self.repository / entry.path
                if candidate.is_symlink():
                    additions = 1
                else:
                    with candidate.open("r", encoding="utf-8") as stream:
                        additions = sum(1 for _ in stream)
            except (OSError, UnicodeDecodeError):
                additions = None
            result.append(GitFile(path=entry.path, status=entry.status, additions=additions, deletions=0, binary=additions is None))
        return result

    def diff(self, scope: str, path: str, *, base: str | None = None, commit: str | None = None, old_path: str | None = None) -> tuple[str, str | None, str | None]:
        relative = Path(path)
        if relative.is_absolute() or ".." in relative.parts:
            raise WorkbenchError("file path must stay inside the repository", code="path_invalid")
        files = self.changed_files(scope, base=base, commit=commit)
        selected = next((item for item in files if item.path == path), None)
        if selected is None:
            raise WorkbenchError("file is not in the selected changes", code="file_not_changed")
        paths = [selected.old_path, path] if selected.old_path else [path]
        if scope == "working":
            left, right = "HEAD", None
            if not self.head() or any(status == "??" and item == path for status, item in self.status_porcelain()):
                command = ["diff", "--no-ext-diff", "--no-index", "--unified=80", "/dev/null", path]
                left = None
            else:
                command = ["diff", "--no-ext-diff", "--unified=80", "HEAD", "--", *paths]
        elif scope == "branch":
            if not base:
                raise WorkbenchError("base is required for branch diff", code="base_missing")
            left, right = base, self.head()
            command = ["diff", "--no-ext-diff", "--unified=80", base, right or "HEAD", "--", *paths]
        elif scope == "commit":
            if not commit:
                raise WorkbenchError("commit is required", code="commit_required")
            commit = self.resolve_commit(commit)
            parent = self.run(["rev-list", "--parents", "-n", "1", commit]).stdout.strip().split()[1:2]
            left, right = (parent[0] if parent else None), commit
            command = (["diff", left, commit] if left else ["diff-tree", "--root", "--no-commit-id", "-r", commit]) + ["--no-ext-diff", "--unified=80", "--", *paths]
        else:
            raise WorkbenchError(f"unsupported diff scope: {scope}", code="scope_invalid")
        result = self.run(command, check=False)
        if result.returncode not in {0, 1}:
            detail = result.stderr.strip() or result.stdout.strip() or "git diff failed"
            raise GitCommandError(detail, code="git_diff_failed", details={"repository": str(self.repository), "path": path})
        return result.stdout, left, right

    def resolve_commit(self, value: str) -> str:
        result = self.run(["rev-parse", "--verify", "--end-of-options", f"{value}^{{commit}}"], check=False)
        if result.returncode:
            raise WorkbenchError("commit is unavailable", code="commit_missing")
        return result.stdout.strip()

    def graph(self, *, mode: str = "branch", base: str | None = None, max_commits: int = 50) -> dict[str, object]:
        head = self.head()
        if not head:
            return {"historyMode": mode, "nodes": [], "refs": [], "hasOlder": False, "loadedCount": 0, "baseLoaded": False}
        limit = max(1, min(int(max_commits), 200))
        args = ["log", "--topo-order", "--date-order", f"--max-count={limit + 1}", "--pretty=format:%H%x00%P%x00%h%x00%s%x00%an%x00%aI"]
        if mode == "branch":
            if not base:
                raise WorkbenchError("base is required for branch graph", code="base_missing")
            args.append(f"{base}..{head}")
        elif mode != "full":
            raise WorkbenchError(f"unsupported graph history mode: {mode}", code="history_mode_invalid")
        output = self.run(args).stdout
        raw_nodes = [line for line in output.splitlines() if line.strip()]
        has_older = len(raw_nodes) > limit
        raw_nodes = raw_nodes[:limit]
        nodes: list[dict[str, object]] = []
        seen: set[str] = set()
        for line in raw_nodes:
            fields = line.split("\x00")
            if len(fields) < 6:
                continue
            sha, parents, short_sha, subject, author, authored_at = fields[:6]
            if not sha or sha in seen:
                continue
            seen.add(sha)
            nodes.append({
                "sha": sha,
                "shortSha": short_sha,
                "parents": [value for value in parents.split() if value],
                "subject": subject,
                "author": author,
                "authoredAt": authored_at,
                "isBase": bool(base and sha == base),
            })
        if mode == "branch" and base and base not in seen:
            base_subject = self.run(["show", "-s", "--format=%s", base], check=False).stdout.strip() or "base"
            nodes.append({"sha": base, "shortSha": base[:8], "parents": [], "subject": base_subject, "author": "", "authoredAt": "", "isBase": True})
            seen.add(base)
        refs = [ref for ref in self.refs(head=head) if str(ref.get("sha")) in seen]
        decorations: dict[str, list[dict[str, object]]] = {}
        for ref in refs:
            decorations.setdefault(str(ref["sha"]), []).append(ref)
        for node in nodes:
            node_refs = decorations.get(str(node["sha"]), [])
            node["refs"] = node_refs
            node["decorations"] = [str(ref.get("shortName") or ref.get("name") or "") for ref in node_refs]
            if node["sha"] == head:
                node["decorations"] = ["HEAD", *[value for value in node["decorations"] if value != "HEAD"]]
            if len(node.get("parents", [])) > 1:
                node["mergeSources"] = [
                    {"parentSha": parent, "refs": decorations.get(str(parent), [])}
                    for parent in node["parents"][1:]
                ]
        return {
            "historyMode": mode,
            "nodes": nodes,
            "refs": refs,
            "hasOlder": has_older,
            "loadedCount": len(nodes) if mode == "full" else len([node for node in nodes if not node.get("isBase")]),
            "baseLoaded": any(bool(node.get("isBase")) for node in nodes),
        }

    def fingerprint(self) -> str:
        if not self.is_repository():
            return "missing"
        head = self.head() or ""
        branch = self.branch() or ""
        upstream, upstream_sha = self.upstream()
        status = "\x00".join(f"{code}:{path}" for code, path in self.status_porcelain())
        refs = "\x00".join(f"{item['name']}:{item['sha']}" for item in self.refs(head=head))
        return hashlib.sha256("\x00".join([head, branch, upstream or "", upstream_sha or "", status, refs]).encode()).hexdigest()

File: paseo-plugin/server/backend/git.ts
import { AsyncLocalStorage } from "node:async_hooks";
import { existsSync, lstatSync, readFileSync, readlinkSync } from "node:fs";
import { join, isAbsolute } from "node:path";
import { canonical, inside, WorkbenchError, type Json } from "./storage.ts";
import { command } from "./process.ts";

export type GitFile = {
  path: string;
  status: string;
  oldPath: string | null;
  additions: number | null;
  deletions: number | null;
  binary: boolean;
};
let running = 0;
const priority = new AsyncLocalStorage<number>();
let commands = 0, timedOut = 0;
export const gitDiagnostics = () => ({ running, queued: waiting.length, commands, timedOut });
export const withBackgroundGit = <T>(operation: () => Promise<T>): Promise<T> => priority.run(1, operation);
type GitWaiter = {
  priority: number;
  resolve: () => void;
  reject: (error: WorkbenchError) => void;
  timer?: ReturnType<typeof setTimeout>;
  signal?: AbortSignal;
  onAbort?: () => void;
  settled: boolean;
};
const waiting: GitWaiter[] = [];
function observationTimeout(): WorkbenchError {
  return new WorkbenchError(
    "observation_timeout",
    "observation deadline exceeded while waiting for a Git slot",
  );
}
function removeWaiter(waiter: GitWaiter): void {
  const index = waiting.indexOf(waiter);
  if (index >= 0) waiting.splice(index, 1);
}
async function acquireGitSlot(deadline?: number, signal?: AbortSignal): Promise<void> {
  if (signal?.aborted)
    throw new WorkbenchError("observer_cancelled", "observation cancelled");
  if (deadline !== undefined && Date.now() >= deadline)
    throw observationTimeout();
  if (running < 4) {
    running++;
    return;
  }
  await new Promise<void>((resolve, reject) => {
    const waiter: GitWaiter = {
      priority: priority.getStore() || 0,
      resolve,
      reject,
      signal,
      settled: false,
    };
    const remaining = deadline === undefined ? undefined : deadline - Date.now();
    if (remaining !== undefined && remaining <= 0) {
      waiter.settled = true;
      reject(observationTimeout());
      return;
    }
    waiting.push(waiter);
    waiting.sort((a, b) => a.priority - b.priority);
    if (remaining !== undefined)
      waiter.timer = setTimeout(() => {
        if (waiter.settled) return;
        removeWaiter(waiter);
        waiter.settled = true;
        waiter.signal?.removeEventListener("abort", waiter.onAbort!);
        waiter.reject(observationTimeout());
      }, remaining);
    const onAbort = () => {
      if (waiter.settled) return;
      removeWaiter(waiter);
      waiter.settled = true;
      if (waiter.timer) clearTimeout(waiter.timer);
      waiter.reject(new WorkbenchError("observer_cancelled", "observation cancelled"));
    };
    waiter.onAbort = onAbort;
    signal?.addEventListener("abort", onAbort, { once: true });
  });
}
function releaseGitSlot(): void {
  while (waiting.length) {
    const waiter = waiting.shift()!;
    if (waiter.settled) continue;
    waiter.settled = true;
    if (waiter.timer) clearTimeout(waiter.timer);
    waiter.signal?.removeEventListener("abort", waiter.onAbort!);
    waiter.resolve();
    return;
  }
  running--;
}
async function gitSlot<T>(operation: () => Promise<T>, deadline?: number, signal?: AbortSignal): Promise<T> {
  let acquired = false;
  await acquireGitSlot(deadline, signal);
  acquired = true;
  try {
    if (deadline !== undefined && Date.now() >= deadline)
      throw observationTimeout();
    return await operation();
  } finally {
    if (acquired) releaseGitSlot();
  }
}
export class Git {
  path: string;
  timeout: number;
  deadline?: number;
  signal?: AbortSignal;
  constructor(path: string, timeout = 3000, deadline?: number, signal?: AbortSignal) {
    this.path = canonical(path);
    this.timeout = timeout;
    this.deadline = deadline;
    this.signal = signal;
  }
  async run(args: string[], check = true) {
    if (this.signal?.aborted) throw new WorkbenchError("observer_cancelled", "Git request cancelled");
    let result;
    try {
      result = await gitSlot(() =>
        {
          const remaining = this.deadline === undefined
            ? this.timeout
            : this.deadline - Date.now();
          if (remaining <= 0) throw observationTimeout();
          commands++;
          return command(
            "git",
            ["-c", "core.fsmonitor=false", "-C", this.path, ...args],
            {
              cwd: this.path,
              timeout: Math.max(1, Math.min(this.timeout, remaining)),
              env: {
                ...process.env,
                GIT_OPTIONAL_LOCKS: "0",
                GIT_TERMINAL_PROMPT: "0",
                GIT_EXTERNAL_DIFF: "",
              },
              signal: this.signal,
            },
          );
        },
        this.deadline,
        this.signal,
      );
      if (this.deadline !== undefined && Date.now() >= this.deadline)
        throw observationTimeout();
    } catch (error) {
      if (error instanceof WorkbenchError && error.code === "observation_timeout")
        throw new WorkbenchError("observation_timeout", error.message, {
          repository: this.path,
          args,
        });
      if (error instanceof WorkbenchError && error.code === "process_timeout") {
        timedOut++;
        throw new WorkbenchError(
          this.deadline !== undefined && Date.now() >= this.deadline
            ? "observation_timeout"
            : "git_timeout",
          error.message,
          { repository: this.path, args },
        );
      }
      throw error;
    }
    if (check && result.code !== 0)
      throw new WorkbenchError(
        "git_command_failed",
        result.stderr.trim() || "Git command failed",
        { repository: this.path, args, returncode: result.code },
      );
    return result;
  }
  async text(args: string[], check = true) {
    return (await this.run(args, check)).stdout.trim();
  }
  async valid() {
    return (
      existsSync(this.path) &&
      (await this.text(["rev-parse", "--is-inside-work-tree"], false)) ===
        "true"
    );
  }
  async root() {
    return canonical(await this.text(["rev-parse", "--show-toplevel"]));
  }
  async head() {
    const r = await this.run(["rev-parse", "--verify", "HEAD"], false);
    return r.code === 0 ? r.stdout.trim() : null;
  }
  async branch() {
    return (
      (await this.text(
        ["symbolic-ref", "--quiet", "--short", "HEAD"],
        false,
      )) || null
    );
  }
  async refsAtHead(head: string): Promise<string[]> {
    if (!head) return [];
    const result = await this.run([
      "for-each-ref",
      "--format=%(refname:short)",
      "--points-at",
      head,
      "refs/heads",
      "refs/remotes",
    ], false);
    if (result.code !== 0) return [];
    return [...new Set(result.stdout.split(/\r?\n/).map((value) => value.trim()).filter(Boolean))];
  }
  async upstream(): Promise<[string | null, string | null]> {
    const name = await this.run(
        ["rev-parse", "--abbrev-ref", "--symbolic-full-name", "@{upstream}"],
        false,
      ),
      sha = await this.run(["rev-parse", "--verify", "@{upstream}"], false);
    return [
      name.code === 0 ? name.stdout.trim() : null,
      sha.code === 0 ? sha.stdout.trim() : null,
    ];
  }
  async commit(value: string) {
    const r = await this.run(
      ["rev-parse", "--verify", "--end-of-options", `${value}^{commit}`],
      false,
    );
    if (r.code)
      throw new WorkbenchError("commit_missing", "commit is unavailable");
    return r.stdout.trim();
  }
  async status(ignoreSubmoduleContent: boolean | "all" = false, includeIgnored = false): Promise<Array<[string, string]>> {
    const values = (
        await this.run([
          "status",
          "--porcelain=v1",
          "-z",
          "--untracked-files=all",
          ...(includeIgnored ? ["--ignored=matching"] : []),
          ...(ignoreSubmoduleContent ? [`--ignore-submodules=${ignoreSubmoduleContent === "all" ? "all" : "dirty"}`] : []),
        ])
      ).stdout.split("\0"),
      result: Array<[string, string]> = [];
    for (let i = 0; i < values.length; i++) {
      const value = values[i];
      if (!value) continue;
      const code = value.slice(0, 2),
        path = value.slice(3);
      result.push([
        code,
        /[RC]/.test(code[0]) && values[i + 1]
          ? `${path}\0${values[++i]}`
          : path,
      ]);
    }
    return result;
  }
  async worktrees(): Promise<Json[]> {
    return (await this.text(["worktree", "list", "--porcelain"]))
      .split(/\n\n/)
      .filter(Boolean)
      .map((entry) =>
        Object.fromEntries(
          entry.split("\n").map((line) => {
            const space = line.indexOf(" ");
            return space < 0
              ? [line, true]
              : [line.slice(0, space), line.slice(space + 1)];
          }),
        ),
      );
  }
  async registered(path: string) {
    return (await this.worktrees()).find(
      (entry) =>
        entry.worktree && canonical(entry.worktree) === canonical(path),
    );
  }
  async completed(path: string, branch: string, sha: string) {
    const entry = await this.registered(path);
    return (
      existsSync(path) &&
      entry &&
      !entry.locked &&
      entry.branch === `refs/heads/${branch}` &&
      entry.HEAD === sha
    );
  }
  async refs(head?: string | null): Promise<Json[]> {
    const lines = (
      await this.text([
        "for-each-ref",
        ...(head ? [`--merged=${head}`] : []),
        "--format=%(objectname)\t%(refname)\t%(symref)\t%(*objectname)",
        "refs/heads",
        "refs/remotes",
        "refs/tags",
      ])
    ).split("\n");
    return lines
      .filter(Boolean)
      .map((line) => {
        const [sha, name, , peeled] = line.split("\t");
        const kind = name.startsWith("refs/tags/")
          ? "tag"
          : name.startsWith("refs/remotes/")
            ? "remote"
            : "local";
        return {
          sha: kind === "tag" && peeled ? peeled : sha,
          name,
          shortName: name.replace(/^refs\/(heads|remotes|tags)\//, ""),
          kind,
        };
      })
      .sort((a, b) =>
        `${a.kind}:${a.shortName}`.localeCompare(`${b.kind}:${b.shortName}`),
      );
  }
  async range(
    scope: string,
    base?: string | null,
    commit?: string | null,
  ): Promise<{ args: string[]; left: string | null; right: string | null }> {
    if (scope === "working")
      return { args: ["diff", "HEAD"], left: "HEAD", right: null };
    if (scope === "branch") {
      if (!base) throw new WorkbenchError("base_missing", "base is required");
      const left = await this.commit(base),
        right = await this.head();
      return { args: ["diff", left, right || "HEAD"], left, right };
    }
    if (scope !== "commit")
      throw new WorkbenchError("scope_invalid", "unsupported scope");
    if (!commit)
      throw new WorkbenchError("commit_required", "commit is required");
    const right = await this.commit(commit),
      parents = (await this.text(["rev-list", "--parents", "-n", "1", right]))
        .split(/\s+/)
        .slice(1);
    return {
      args: parents.length
        ? ["diff", parents[0], right]
        : ["diff-tree", "--root", "--no-commit-id", "-r", right],
      left: parents[0] || null,
      right,
    };
  }
  async files(
    scope: string,
    base?: string | null,
    commit?: string | null,
    ignoreSubmoduleContent = false,
  ): Promise<GitFile[]> {
    const hasHead = await this.head(),
      range = await this.range(scope, base, commit);
    const diffArgs = scope === "working" && ignoreSubmoduleContent
      ? [...range.args, "--ignore-submodules=dirty"] : range.args;
    const empty = scope === "working" && !hasHead;
    const names = empty
      ? []
      : (
          await this.run([
            ...diffArgs,
            "--no-ext-diff",
            "--name-status",
            "-z",
            "--find-renames",
          ])
        ).stdout.split("\0");
    const result: GitFile[] = [];
    for (let i = 0; i < names.length; ) {
      const status = names[i++];
      if (!status) continue;
      let path = names[i++],
        oldPath: string | null = null;
      if (/^[RC]/.test(status)) {
        oldPath = path;
        path = names[i++];
      }
      result.push({
        path,
        status: status[0],
        oldPath,
        additions: null,
        deletions: null,
        binary: false,
      });
    }
    const stats = empty
      ? []
      : (
          await this.run([...diffArgs, "--no-ext-diff", "--numstat", "-z"])
        ).stdout.split("\0");
    for (let i = 0; i < stats.length; ) {
      const fields = stats[i++].split("\t");
      if (fields.length < 3) continue;
      let path = fields.slice(2).join("\t");
      if (!path) {
        i++;
        path = stats[i++];
      }
      const file = result.find((item) => item.path === path);
      if (file)
        Object.assign(file, {
          additions: fields[0] === "-" ? null : Number(fields[0]),
          deletions: fields[1] === "-" ? null : Number(fields[1]),
          binary: fields[0] === "-",
        });
    }
    if (scope === "working")
      for (const [status, path] of await this.status())
        if (status === "??" && !result.some((item) => item.path === path)) {
          let additions: number | null = null;
          try {
            const target = join(this.path, path);
            if (!inside(target, this.path)) throw new Error("outside");
            const data = lstatSync(target).isSymbolicLink()
              ? Buffer.from(readlinkSync(target))
              : readFileSync(target);
            const text = new TextDecoder("utf-8", { fatal: true }).decode(data);
            if (!data.includes(0))
              additions = text
                ? text.split("\n").length - (text.endsWith("\n") ? 1 : 0)
                : 0;
          } catch {}
          result.push({
            path,
            status: "A",
            oldPath: null,
            additions,
            deletions: 0,
            binary: additions === null,
          });
        }
    return result;
  }
  async diff(
    scope: string,
    path: string,
    base?: string | null,
    commit?: string | null,
  ) {
    if (isAbsolute(path) || path.split("/").includes(".."))
      throw new WorkbenchError("path_invalid", "diff path must be relative");
    const file = (await this.files(scope, base, commit)).find(
      (item) => item.path === path,
    );
    if (!file)
      throw new WorkbenchError(
        "file_not_changed",
        "file is not in selected changes",
      );
    const range = await this.range(scope, base, commit);
    const untracked =
      scope === "working" &&
      (!(await this.head()) ||
        (await this.status()).some(
          ([code, name]) => code === "??" && name === path,
        ));
    if (untracked && !inside(join(this.path, path), this.path))
      throw new WorkbenchError(
        "path_invalid",
        "untracked file points outside repository",
      );
    const args = untracked
      ? [
          "diff",
          "--no-ext-diff",
          "--no-index",
          "--unified=80",
          "--",
          "/dev/null",
          path,
        ]
      : [
          ...range.args,
          "--no-ext-diff",
          "--unified=80",
          "--",
          ...(file.oldPath ? [file.oldPath] : []),
          path,
        ];
    const result = await this.run(args, false);
    if (result.code !== 0 && !(untracked && result.code === 1))
      throw new WorkbenchError("git_diff_failed", result.stderr);
    return {
      patch: result.stdout,
      left: untracked ? null : range.left,
      right: range.right,
    };
  }
  async graph(mode: string, base: string | null, limit: number) {
    const head = await this.head();
    if (!head)
      return {
        nodes: [],
        refs: [],
        hasOlder: false,
        loadedCount: 0,
        baseLoaded: false,
        historyMode: mode,
      };
    if (!["branch", "full"].includes(mode))
      throw new WorkbenchError("history_mode_invalid", "invalid history mode");
    if (mode === "branch" && !base)
      throw new WorkbenchError("base_missing", "base is required");
    const safeBase = base ? await this.commit(base) : null;
    const lines = (
      await this.text([
        "log",
        "--topo-order",
        "--date-order",
        `--max-count=${limit + 1}`,
        "--pretty=format:%H%x00%P%x00%h%x00%s%x00%an%x00%aI",
        ...(mode === "branch" ? [`${safeBase}..${head}`] : []),
      ])
    )
      .split("\n")
      .filter(Boolean);
    const nodes: Json[] = lines.slice(0, limit).map((line) => {
      const [sha, parents, shortSha, subject, author, authoredAt] =
        line.split("\0");
      return {
        sha,
        parents: parents.split(" ").filter(Boolean),
        shortSha,
        subject,
        author,
        authoredAt,
        isBase: sha === safeBase,
      };
    });
    if (
      mode === "branch" &&
      safeBase &&
      !nodes.some((node) => node.sha === safeBase)
    )
      nodes.push({
        sha: safeBase,
        shortSha: safeBase.slice(0, 8),
        parents: [],
        subject:
          (await this.text(["show", "-s", "--format=%s", safeBase], false)) ||
          "base",
        author: "",
        authoredAt: "",
        isBase: true,
      });
    const refs = (await this.refs(head)).filter((ref) =>
      nodes.some((node) => node.sha === ref.sha),
    );
    const branch = await this.branch(),
      [upstream] = await this.upstream();
    refs.forEach((ref) =>
      Object.assign(ref, {
        isHead: ref.kind === "local" && ref.shortName === branch,
        isUpstream: ref.shortName === upstream,
      }),
    );
    nodes.forEach((node) => {
      node.refs = refs.filter((ref) => ref.sha === node.sha);
      node.decorations = [
        ...(node.sha === head ? ["HEAD"] : []),
        ...node.refs.map((ref: Json) => ref.shortName),
      ];
      if (node.parents.length > 1)
        node.mergeSources = node.parents
          .slice(1)
          .map((parentSha: string) => ({
            parentSha,
            refs: refs.filter((ref) => ref.sha === parentSha),
          }));
    });
    return {
      nodes,
      refs,
      hasOlder: lines.length > limit,
      loadedCount:
        mode === "full"
          ? nodes.length
          : nodes.filter((node) => !node.isBase).length,
      baseLoaded: nodes.some((node) => node.isBase),
      historyMode: mode,
    };
  }
}
